### PR TITLE
Support function translation to Databricks SQL in TSql and Snowflake

### PR DESCRIPTION
**Pull request by jimidle**
Some functions must be translated from TSQL or Snowflake versions into the equivalent IR for Databricks SQL. In some cases the function must be translated in one dialect, say TSql but is equivalent in another, say Snowflake.

Here we upgrade the FunctionBuilder system to be dialect aware, and provide a ConversionStrategy system that allows for any type of conversion from a simple name translation or more complicated specific IR representations when there is no equivalent.

For example the TSQL code
```tsql
SELECT ISNULL(x, 0)
```

Should translate to:

```sql
SELECT IFNULL(x, 0)
```

In Databricks SQL, but in Snowflake SQL:

```snowflake
SELECT ISNULL(col)
```

Is directly equivalent to Databricks SQL and needs no conversion.

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionConverters.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  // All bets are off if the original name was mixed case, but that is rarely seen in SQL and we are
+  // just making reasonable efforts here.
+  private def convertString(irName: String, newName: String): String = {
+    if (irName.forall(_.isLower)) newName.toLowerCase(Locale.ROOT) else newName
+  }
+
+  object FunctionRename extends ConversionStrategy {
+    override def convert(irName: String, args: Seq[ir.Expression], dialect: SqlDialect): ir.Expression = {
+      (irName.toUpperCase(), dialect) match {
+        case ("ISNULL", TSql) => ir.CallFunction(FunctionConverters.convertString(irName, "IFNULL"), args)
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionConverters.scala:26]: A quick search into TSQL docs didn't yield any result for `IFNULL`, did I miss something?

File: core/src/test/scala/com/databricks/labs/remorph/parsers/common/FunctionBuilderSpec.scala [hunk 2]
```diff
@@ -283,39 +282,71 @@ class FunctionBuilderSpec extends AnyFlatSpec with Matchers with TableDrivenProp
...
-    val result5 = FunctionBuilder.buildFunction("'nonmatching", Seq())
+    val result5 = FunctionBuilder.buildFunction("'nonmatching", List.empty, TSql)
     result5 match {
-      case f: UnresolvedFunction => f.function_name shouldBe "'nonmatching"
+      case f: ir.UnresolvedFunction => f.function_name shouldBe "'nonmatching"
       case _ => fail("Unexpected function type")
     }
   }
 
+  "buildFunction" should "Apply known TSQL conversion strategies" in {
```
Comment (vil1) [inline core/src/test/scala/com/databricks/labs/remorph/parsers/common/FunctionBuilderSpec.scala:320]: as a side note, we could have a single block for tests around `buildFunction` like in:
```scala
"buildFunction" should {
  "Apply known TSQL conversion strategies" in {
    // ...
  }
  // etc.
}
```

File: core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionBuilder.scala [hunk 3]
```diff
@@ -142,7 +148,9 @@ object FunctionBuilder {
     case "ISDATE" => Some(FixedArity(1))
     case "ISDESCENDANTOF" => Some(FixedArity(1))
     case "ISJSON" => Some(VariableArity(1, 2))
-    case "ISNULL" => Some(FixedArity(2))
+    // Though ISNULL is FixedArity, it is one argument in Snowflake (and others) but two arguments in TSql
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionBuilder.scala:151]: This difference isn't addressed by the code bellow (we get a `VariableArity` even in the case of Snowflake).

**Review by vil1 — COMMENTED** [review]
I suggest changes to this PR in #418

File: core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionConverters.scala (hunk 1 above)
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionConverters.scala:26]: Yes. TSQL uses ISNULL(Col, <default>) but the funciotn in ANSO SQL and Databricks is IFNULL(col, <default) and uses ISNULL(col) to return true/false.

**Review by vil1 — APPROVED** [review]
LGTM

**Review by nfx — CHANGES_REQUESTED** [review]

File: core/src/main/scala/com/databricks/labs/remorph/parsers/ParserCommon.scala [hunk 4]
```diff
@@ -7,3 +7,7 @@ trait ParserCommon {
     a != null && b != null && a.getSourceInterval.startsBeforeDisjoint(b.getSourceInterval)
   }
 }
+
+sealed trait SqlDialect
```
[the remark this answers is not in the document]
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/ParserCommon.scala:14]: OK

File: core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionBuilder.scala [hunk 5]
```diff
@@ -2,277 +2,286 @@ package com.databricks.labs.remorph.parsers
...
-    case "USER_ID" => Some(VariableArity(0, 1))
-    case "USER_NAME" => Some(VariableArity(0, 1))
-    case "VALUE" => Some(FixedArity(2, XmlFunction))
-    case "VAR" => Some(FixedArity(1))
-    case "VARP" => Some(FixedArity(1))
-    case "XACT_STATE" => Some(FixedArity(0))
-    case "YEAR" => Some(FixedArity(1))
-    case _ => None
+  private val functionDefinitionPf: PartialFunction[(SqlDialect, String), FunctionDefinition] = {
+    case (_, "ABS") => FunctionDefinition.standard(1)
```
[the remark this answers is not in the document]
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/FunctionBuilder.scala]: For now, we have various ConversionStrategy objects, but as some conversions will be intricate, then a class approach may well be better.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/ParserCommon.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/ParserCommon.scala]: This has now been completely removed.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala [hunk 6]
```diff
@@ -281,7 +280,7 @@ class TSqlExpressionBuilder extends TSqlParserBaseVisitor[ir.Expression] with Pa
   override def visitStandardFunction(ctx: StandardFunctionContext): ir.Expression = {
     val name = ctx.funcId.getText
     val args = Option(ctx.expression()).map(_.asScala.map(_.accept(this))).getOrElse(Seq.empty)
-    FunctionBuilder.buildFunction(name, args)
+    FunctionBuilder.buildFunction(name, args, TSql)
```
[the remark this answers is not in the document]
Comment (jimidle) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlExpressionBuilder.scala]: Reimplemented as requested.

**Review by vil1 — APPROVED** [review]
LGTM

**Review by nfx — APPROVED** [review]
Lgtm

**Pull request merged**